cmd/kubectl-gadget/top/ebpf.go

```diff
@@ -0,0 +1,296 @@
+				types.SortByParam:   sortBy,
+			},
+		}
+
+		// when params.Timeout == interval it means the user
+		// only wants to run for a given amount of time and print
+		// that result.
+		singleShot := params.Timeout == outputInterval
+
+		// start print loop if this is not a "single shoot" operation
+		if singleShot {
+			ebpfPrintHeader()
+		} else {
+			ebpfStartPrintLoop()
+		}
+
+		if err := utils.RunTraceStreamCallback(config, ebpfCallback); err != nil {
+			return commonutils.WrapInErrRunGadget(err)
+		}
+
+		if singleShot {
+			ebpfPrintEvents()
+		}
+
+		return nil
+	},
+	SilenceUsage: true,
+	PreRunE: func(cmd *cobra.Command, args []string) error {
+		var err error
+		ebpfSortBy, err = types.ParseSortBy(sortBy)
+		if err != nil {
+			return commonutils.WrapInErrInvalidArg("--sort", err)
+		}
+
+		return nil
+	},
+	Args: cobra.MaximumNArgs(1),
+}
+
+func init() {
+	addTopCommand(ebpfCmd, types.MaxRowsDefault, types.SortBySlice)
+}
+
+func ebpfCallback(line string, node string) {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	var event types.Event
+
+	if err := json.Unmarshal([]byte(line), &event); err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %s", commonutils.WrapInErrUnmarshalOutput(err, line))
+		return
+	}
+
+	if event.Error != "" {
+		fmt.Fprintf(os.Stderr, "Error: failed on node %q: %s", event.Node, event.Error)
+		return
+	}
+
+	ebpfNodeStats[node] = event.Stats
+}
+
+func ebpfStartPrintLoop() {
+	go func() {
+		ticker := time.NewTicker(time.Duration(outputInterval) * time.Second)
+		ebpfPrintHeader()
+		for {
+			_ = <-ticker.C
+			ebpfPrintHeader()
+			ebpfPrintEvents()
+		}
+	}()
+}
+
+func ebpfPrintHeader() {
+	switch params.OutputMode {
+	case commonutils.OutputModeColumns:
+		if term.IsTerminal(int(os.Stdout.Fd())) {
+			utils.ClearScreen()
+		} else {
+			fmt.Println("")
+		}
+		fmt.Printf("%-16s %-8s %-16s %-16s %-7s %-20s %-12s %-6s\n",
+			"NODE",
+			"PROG",
+			"TYPE",
+			"NAME",
+			"PID", "COMM",
+			"TIME",
+			"COUNT",
+		)
+	case commonutils.OutputModeCustomColumns:
+		if term.IsTerminal(int(os.Stdout.Fd())) {
+			utils.ClearScreen()
+		} else {
+			fmt.Println("")
+		}
+		fmt.Println(ebpfGetCustomColsHeader(params.CustomColumns))
+	}
+}
+
+func ebpfPrintEvents() {
+	// sort and print events
+	mutex.Lock()
+
+	stats := []types.Stats{}
+	for _, stat := range ebpfNodeStats {
+		stats = append(stats, stat...)
+	}
+	ebpfNodeStats = make(map[string][]types.Stats)
+
+	mutex.Unlock()
+
+	types.SortStats(stats, ebpfSortBy)
+
+	switch params.OutputMode {
+	case commonutils.OutputModeColumns:
+		for idx, event := range stats {
+			if idx == maxRows {
+				break
+			}
+			pid := ""
+			comm := ""
+			if len(event.Pids) > 0 {
+				pid = strconv.Itoa(int(event.Pids[0].Pid))
+				comm = event.Pids[0].Comm
+			}
+			fmt.Printf("%-16s %-8d %-16s %-16s %-7s %-20s %12v %6d\n",
+				event.Node,
+				event.ProgramID,
+				event.Type,
+				event.Name,
+				pid, comm,
+				time.Duration(event.Runtime),
+				event.RunCount,
```

Comment:
Is there any way to align the time to the left? It looks wear:

```bash
NODE             PROG     TYPE             NAME             PID     COMM                 TIME         COUNT
master           32       Tracing          gadget_ebpftop   129449  gadgettracerman           37.18µs   1447
master           11       CGroupSKB                         1       systemd                        0s      0
master           15       CGroupSKB                         1       systemd                        0s      0
master           16       CGroupSKB                         1       systemd                        0s      0

```

Reply:
Hmm, I think aligning the time to the right makes it much more readable/comparable (especially when you sort by time), same goes for the count.

Specifically regarding the time: go will use whatever time unit suits best with %v, so sometimes the suffix will be "s", sometimes it's "µs" (like in your example) and so on. That makes it even more tricky to comparable it to other rows. However, not using the unit automation (always using "ms" for example) could lead to massive or always rounded to zero numbers.

Unsorted it would look something like that (left vs right aligned):
```
TIME         TIME
1.13ms     1.13ms
1.05s       1.05s
10ms         10ms
8s             8s
```

Sorted:
```
TIME         TIME
8s             8s
1.05s       1.05s
10ms         10ms
1.13ms     1.13ms
```

I would always prefer the right aligned values, IMHO.

Maybe the header should be aligned to the right as well? What do you think?

Reply:
You're right! I think aligning the columns in the same way would improve the view.